### PR TITLE
fix: List UIViewControllers sub classes

### PR DESCRIPTION
**Pull request by brustolin**
## :pencil: Checklist

- [ ] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [ ] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentrySubClassFinder.m [hunk 1]
```diff
@@ -135,4 +120,55 @@ - (int)getClassList:(Class **)classes
...
+                             andLevel:kSentryLevelDebug];
+            return;
+        }
+
+        NSMutableArray<NSString *> *classesToSwizzle = [NSMutableArray new];
+        for (int i = 0; i < count; i++) {
+            NSString *className = [NSString stringWithUTF8String:classes[i]];
+            if ([className hasSuffix:@"ViewController"]) {
+                Class class = NSClassFromString(className);
+                if ([self isClass:class subClassOf:viewControllerClass]) {
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:159]: `h`: We cant call this method on a background thread because of
https://github.com/getsentry/sentry-cocoa/blob/cb117ceaf5bb343a72635e3079269185a75b09b8/Sources/Sentry/SentrySubClassFinder.m#L61-L68

Maybe `objc_copyClassNamesForImage` is fast enough for running it on the main thread. If not we could just call `objc_copyClassNamesForImage` and then switch to the main thread.

**Comment by philipphofmann** [comment]
Please also bump react-native after merging this, as it seems like this is also causing crashes there https://github.com/getsentry/sentry-react-native/issues/1964.

File: Sources/Sentry/SentrySubClassFinder.m (hunk 1 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:159]: @philipphofmann  according to my tests, this solution does not trigger the class initializer.
The method `isClass:subClassOf:` is a local method, not a `Class` method.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ## 7.10.0
 
+- fix: Swizzling UIViewControllers crash (#1670)
```
Comment (bruno-garcia) [inline CHANGELOG.md:5]: Needs to move to the Unreleased block

File: Sources/Sentry/SentrySubClassFinder.m [hunk 3]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
...
-            if (superClass) {
-                [indexesToSwizzle addObject:@(i)];
+        // which assume they are running on the main thread. Therefore, we store the class name
+        // instead so we can search for the subclasses on a background thread. We can't use
+        // NSObject:isSubclassOfClass as not all classes in the runtime in classes inherit from
+        // NSObject and a call to isSubclassOfClass would call the initializer of the class, which
+        // we can't allow because of the problem with UIViewControllers mentioned above.
+        NSMutableArray<NSString *> *classesToSwizzle = [NSMutableArray new];
+        for (int i = 0; i < count; i++) {
+            NSString *className = [NSString stringWithUTF8String:classes[i]];
```
Comment (bruno-garcia) [inline Sources/Sentry/SentrySubClassFinder.m:61]: potentially configurable?

**Review by bruno-garcia — APPROVED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (brustolin) [reply CHANGELOG.md:5]: Merging problem. 

Fixed

File: Sources/Sentry/SentrySubClassFinder.m (hunk 3 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:61]: Yes. I would like to have an option that allows the user to choose the classes to be swizzled.

Creating an issue for this.

File: Sources/Sentry/SentryDefaultObjCRuntimeWrapper.m [hunk 4]
```diff
@@ -9,6 +9,11 @@ - (int)getClassList:(__unsafe_unretained Class *)buffer bufferCount:(int)bufferC
     return objc_getClassList(buffer, bufferCount);
```
Comment (philipphofmann) [inline Sources/Sentry/SentryDefaultObjCRuntimeWrapper.m:9]: `m`: Can't we remove `getClassList`?

File: Sources/Sentry/SentrySubClassFinder.m [hunk 5]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
...
+            if ([className hasSuffix:@"ViewController"]) {
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:62]: `m`: Shouldn't we maybe go with `containsString` as we anyways check if the class is a sub-class of UIViewController?

File: Sources/Sentry/SentrySubClassFinder.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentrySubClassFinder.m:61]: Should we go with an option for suffix or containsString or a full list of class names?

File: Sources/Sentry/SentrySubClassFinder.m [hunk 6]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
...
-            }
-
-            if (superClass) {
-                [indexesToSwizzle addObject:@(i)];
+        // which assume they are running on the main thread. Therefore, we store the class name
+        // instead so we can search for the subclasses on a background thread. We can't use
+        // NSObject:isSubclassOfClass as not all classes in the runtime in classes inherit from
+        // NSObject and a call to isSubclassOfClass would call the initializer of the class, which
+        // we can't allow because of the problem with UIViewControllers mentioned above.
+        NSMutableArray<NSString *> *classesToSwizzle = [NSMutableArray new];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:59]: `l`: I wondered whether Class calls the initializer or not. 
```suggestion
        // Worth noting that the usage of Class doesn't call the initializer.
        NSMutableArray<NSString *> *classesToSwizzle = [NSMutableArray new];
```

File: Sources/Sentry/SentrySubClassFinder.m [hunk 7]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
...
         }
 
         [self.dispatchQueue dispatchOnMainQueue:^{
-            for (NSNumber *i in indexesToSwizzle) {
-                NSInteger index = [i integerValue];
-                block(classes[index]);
+            for (NSString *className in classesToSwizzle) {
+                block(NSClassFromString(className));
             }
             free(classes);
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:74]: `m`: We can move this to line 69 as we don't use the classes anymore in here.

File: Sources/Sentry/SentrySubClassFinder.m [hunk 8]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
     return self;
 }
 
-- (void)actOnSubclassesOf:(Class)parentClass block:(void (^)(Class))block
+- (void)actOnSubclassesOfViewControllerInImage:(NSString *)imageName block:(void (^)(Class))block;
 {
     [self.dispatchQueue dispatchAsyncWithBlock:^{
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:30]: `m`: Do we know how fast this is now? Maybe we don't need to move this to a background thread anymore?

File: Sources/Sentry/SentryUIViewControllerSwizzling.m [hunk 9]
```diff
@@ -47,84 +47,104 @@ - (instancetype)initWithOptions:(SentryOptions *)options
...
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wshadow"
+
 - (void)start
 {
-    [self swizzleRootViewController];
+    id<SentryUIApplication> app = [self findApp];
+    if (app != nil) {
+        /**
+         * If an app targets, for example, iOS 13 or lower, the UIKit inits the initial/root view
```
Comment (philipphofmann) [inline Sources/Sentry/SentryUIViewControllerSwizzling.m:60]: `l`: If the comment is inline we usually use `//`

File: Sources/Sentry/SentryUIViewControllerSwizzling.m [hunk 10]
```diff
@@ -47,84 +47,104 @@ - (instancetype)initWithOptions:(SentryOptions *)options
...
- * auto-generated transactions for the initial view controller. As we use
- * SentrySwizzleModeOncePerClassAndSuperclasses, we don't have to worry about swizzling twice. We
- * could also use objc_getClassList to lookup sub classes of UIViewController, but the lookup can
- * take around 60ms, which is not acceptable.
- */
-- (void)swizzleRootViewController
+- (id<SentryUIApplication>)findApp
 {
     if (![UIApplication respondsToSelector:@selector(sharedApplication)]) {
         NSString *message = @"UIViewControllerSwizziling: UIApplication doesn't respont to "
```
Comment (philipphofmann) [inline Sources/Sentry/SentryUIViewControllerSwizzling.m:99]: `l`: Just found a small typo that I added.
```suggestion
        NSString *message = @"UIViewControllerSwizziling: UIApplication doesn't respond to "
```

File: Sources/Sentry/SentrySubClassFinder.m [hunk 11]
```diff
@@ -24,115 +25,73 @@ - (instancetype)initWithDispatchQueue:(SentryDispatchQueueWrapper *)dispatchQueu
...
+            copyClassNamesForImage:[imageName cStringUsingEncoding:NSUTF8StringEncoding]
+                            amount:&count];
+#if TEST
         // Only for testing. We want to know in tests if the code iterated over the classes, because
         // iterating in edge cases could lead to crashses. Ideally, we would wrap
         // class_getSuperclass in the SentryObjCRuntimeWrapper and count its invocations. As
         // class_getSuperclass is called in a tight loop doing so would slow down the code
         // significantly. This is pragmatic workaround to find out in tests if the code iterated
         // over the classes.
         [self.objcRuntimeWrapper countIterateClasses];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySubClassFinder.m:49]: `m`: I think we can remove this and also the implementation of `countIterateClasses` now because it's not needed anymore with the current implementation. Previously, I needed a way to validate how often we iterated over the classes.

File: Tests/SentryTests/Integrations/Performance/SentrySubClassFinderTests.swift [hunk 12]
```diff
@@ -1,102 +1,74 @@
...
-        var invocations = -1
-        fixture.runtimeWrapper.numClasses = { numClasses in
-            invocations += 1
-            return numClasses - Int32(invocations)
-        }
-        
-        testActOnSubclassesOf(Child1.self, expected: [])
-        XCTAssertEqual(0, fixture.runtimeWrapper.iterateClassesInvocations)
+    func testActOnSubclassesOfViewController_NoViewController() {
+        fixture.runtimeWrapper.classesNames = { _ in [] }
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/SentrySubClassFinderTests.swift:45]: `h`: You asked me how to write tests for the current implementation. I think you almost have it. The testActOnSubclassesOfViewController already makes sure that everything is called on the proper threads and also validates if the block is called on the right classes. I think we just need to pass in some class names to the runtime wrapper here and then adapt the expected below. Do this for all kinds of edge cases and we are good.

**Review by philipphofmann — COMMENTED** [review]
This looks great. I hope this is going to help us to avoid future crashes and this should also speed up the implementation. 👏

File: Sources/Sentry/SentrySubClassFinder.m (hunk 3 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:61]: I would like a callback. But maybe in combination with one of this other options as fallback. List of class names sounds nice.

But I would add this later and continue this conversation at https://github.com/getsentry/sentry-cocoa/issues/1677

File: Sources/Sentry/SentrySubClassFinder.m (hunk 5 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:62]: Sure, don't see why not.

File: Sources/Sentry/SentrySubClassFinder.m (hunk 6 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:59]: It looks like is not calling the initializer

File: Sources/Sentry/SentrySubClassFinder.m (hunk 8 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:30]: we need to test this in a project with a couple hundred classes.

Comment (philipphofmann) [reply Sources/Sentry/SentrySubClassFinder.m:30]: You can use https://github.com/getsentry/sentry-cocoa/blob/master/scripts/generate-classes.sh for that.

**Review by philipphofmann — COMMENTED** [review]
Already looks better. Thanks 👏

File: Sources/Sentry/SentrySubClassFinder.m (hunk 8 above)
Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:30]: Well, for a project with 1007 View controllers, it took ~140ms in debug to find and swizzle all classes.
But only 10ms on the "find" part, and 70ms on the swizzling part, the rest is because we queue the swizzling part to run on the main thread.

With 100 view controller, the total is 70ms, 5ms to find and 3ms to swizzle.

For comparison, the previous approach, with the same 100 view controllers, took ~90ms total, 12ms to find, 11ms to swizzle.

**Review by philipphofmann — COMMENTED** [review]
Nice to see you deleting so much code.

File: Sources/Sentry/SentrySubClassFinder.m (hunk 8 above)
Comment (philipphofmann) [reply Sources/Sentry/SentrySubClassFinder.m:30]: We have to swizzle on the main thread because swizzling on a background thread can lead to crashes, see GH-1355.

With your discoveries which approach do you think works best?

Comment (brustolin) [reply Sources/Sentry/SentrySubClassFinder.m:30]: Let's keep it the way it is.
At the end of the run cycle, it doesn't make much difference.

**Comment by bguidolim** [comment]
Question: if my classes, which are UIViewController, but don't use the name "ViewController" in the end, won't be detected?
